File: maxprime.erl
-module(maxprime).


%-export([get/1]).
-compile(export_all).


%% get the max prime
%%Num < 3317044064679887385961981
get(Num) ->
	Answer = xmr_test(Num),
	if
		Answer =/= true ->
			FirstGuess = first_guess(Num),
			if
				FirstGuess =/= Num ->
					UpTest = up_test(FirstGuess, Num),
					if
						UpTest =/= [] -> UpTest;
						true -> down_test(FirstGuess)
					end;
				true -> down_test(FirstGuess)
			end;
		true -> Num
	end.


%% 求1~Num 间所有素数的和
sum(Num) ->
	sum_main(lists:seq(3, Num), 0).


sum_main([], SUM) ->
	SUM;
sum_main([H | T], SUM) ->
	Answer = xmr_test(H),
	if
		Answer == true ->
			SUM1 = SUM + H,
			sum_main(T, SUM1);
		true ->
			sum_main(T, SUM)
	end.


%%%%%↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑%%%%%
%%%%%↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑ YOU NEED CARE ABOUT ↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑%%%%%
%%%%%                                                                                                    %%%%%
%%%%%                                                                                                    %%%%%
%%%%%↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓ Internal Fuctions ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓%%%%%
%%%%%↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓%%%%%


xmr_test(Num) ->
	if
		Num < 2047 ->
			mr_test(2, Num);
		Num < 1373653 ->
			(mr_test(2, Num)) andalso (mr_test(3, Num));
		Num < 9080191 ->
			(mr_test(31, Num)) andalso (mr_test(73, Num));
		Num < 25326001 ->
			(mr_test(2, Num)) andalso (mr_test(3, Num)) andalso (mr_test(5, Num));
		Num < 4759123141 ->
			(mr_test(2, Num)) andalso (mr_test(7, Num)) andalso (mr_test(61, Num));
		Num < 2152302898747 ->
			(mr_test(2, Num)) andalso (mr_test(3, Num)) andalso (mr_test(5, Num))
				andalso (mr_test(7, Num)) andalso (mr_test(11, Num));
		Num < 3474749660383 ->
			(mr_test(2, Num)) andalso (mr_test(3, Num)) andalso (mr_test(5, Num))
				andalso (mr_test(7, Num)) andalso (mr_test(11, Num))
				andalso (mr_test(13, Num));
		Num < 341550071728321 ->
			(mr_test(2, Num)) andalso (mr_test(3, Num)) andalso (mr_test(5, Num))
				andalso (mr_test(7, Num)) andalso (mr_test(11, Num))
				andalso (mr_test(13, Num)) andalso (mr_test(17, Num));
		Num < 3825123056546413051 ->
			(mr_test(2, Num)) andalso (mr_test(3, Num)) andalso (mr_test(5, Num))
				andalso (mr_test(7, Num)) andalso (mr_test(11, Num))
				andalso (mr_test(13, Num)) andalso (mr_test(17, Num))
				andalso (mr_test(19, Num)) andalso (mr_test(23, Num));
		Num < 318665857834031151167461 ->
			(mr_test(2, Num)) andalso (mr_test(3, Num))
				andalso (mr_test(5, Num)) andalso (mr_test(7, Num))
				andalso (mr_test(11, Num)) andalso (mr_test(13, Num))
				andalso (mr_test(17, Num)) andalso (mr_test(19, Num))
				andalso (mr_test(23, Num)) andalso (mr_test(29, Num))
				andalso (mr_test(31, Num)) andalso (mr_test(37, Num));
		true ->
			(mr_test(2, Num)) andalso (mr_test(3, Num)) andalso (mr_test(5, Num))
				andalso (mr_test(7, Num)) andalso (mr_test(11, Num)) andalso (mr_test(13, Num))
				andalso (mr_test(17, Num)) andalso (mr_test(19, Num)) andalso (mr_test(23, Num))
				andalso (mr_test(29, Num)) andalso (mr_test(31, Num)) andalso (mr_test(37, Num))
				andalso (mr_test(41, Num))
	end.



mr_test(BaseNum, Num) ->
	PassOrNot = even(Num),
	if
		PassOrNot == false ->
			NumMin1 = Num - 1,
			ModMon = mon(BaseNum, NumMin1, Num),
			if
				ModMon == 1 -> mr_test_main(BaseNum, Num, NumMin1);
				true -> false
			end;
		true -> false
	end.


mon(BaseNum, NumMin1, Num) ->
	BaseNum2 = BaseNum rem Num,
	ModMon = 1,
	mon_main(NumMin1, ModMon, BaseNum2, Num).


mon_main(NumMin1, ModMon, BaseNum2, Num) ->
	if
		NumMin1 =/= 0 ->
			Num_left = NumMin1 div 2,
			Num0 = NumMin1 band 1,
			if
				Num0 == 0 ->
					mon_main(Num_left, ModMon, (BaseNum2 * BaseNum2) rem Num, Num);
				true -> mon_main(Num_left, (ModMon * BaseNum2) rem Num, (BaseNum2 * BaseNum2) rem Num, Num)
			end;
		true ->
			ModMon rem Num
	end.


mr_test_main(BaseNum, Num, NumMin1) ->
	Answer = even(Num),
	if
		Answer == true -> mr_test_even(BaseNum, Num, NumMin1);
		true -> mr_test_odd(BaseNum, Num, NumMin1)
	end.


even(Num) ->
	Num band 1 == 0.


mr_test_even(BaseNum, Num, NumMin1) ->
	NumHalf = NumMin1 div 2,
	ModMon = mon(BaseNum, NumHalf, Num),
	if
		ModMon == NumMin1 -> true;
		ModMon =/= 1 -> fasle;
		true -> mr_test_main(BaseNum, Num, NumHalf)
	end.


mr_test_odd(BaseNum, Num, NumMin1) ->
	ModMon2 = mon(BaseNum, NumMin1, Num),
	if
		(ModMon2 == NumMin1) or (ModMon2 == 1) -> true;
		true -> false
	end.


first_guess(Num) ->
	P_guess = p_guess(Num),
	Num - round(P_guess).


p_guess(Num) ->
	LnNum = math:log(Num),
	LnNum / (1 + (1 / LnNum)
		+ (2 / (math:pow(LnNum, 2)))
		+ (6 / (math:pow(LnNum, 3)))
		+ (24 / (math:pow(LnNum, 4)))
		+ (120 / (math:pow(LnNum, 5)))
		+ (720 / (math:pow(LnNum, 6)))
		+ (5040 / (math:pow(LnNum, 7)))
		+ (40320 / (math:pow(LnNum, 8)))).


down_test(FirstGuess) ->
	FirstGuessMin1 = FirstGuess - 1,
	Answer = xmr_test(FirstGuessMin1),
	if
		Answer == true -> FirstGuessMin1;
		true -> down_test(FirstGuessMin1)
	end.


up_test(FirstGuess, Num) ->
	ThreadNum = erlang:system_info(schedulers_online),
	GapNum = Num - FirstGuess,
	RunTime = GapNum div ThreadNum,
	FirstNum = RunTime * ThreadNum,
	LastNum = GapNum - FirstNum,
	if
		RunTime == 0 ->
			PrimeList = get_prime_list(FirstGuess, GapNum),
			if
				PrimeList =/= [] -> lists:last(lists:sort(PrimeList));
				true -> []
			end;
		true ->
			FirstList = get_prime_list_main(FirstGuess, lists:seq(0, RunTime - 1),
				ThreadNum, []),
			SecondList = get_prime_list(FirstGuess + FirstNum, LastNum),
			PrimeList = FirstList ++ SecondList,
			if
				PrimeList =/= [] -> lists:last(lists:sort(PrimeList));
				true -> []
			end
	end.


get_prime_list(FirstGuess, ThreadNum) ->
	upmap(fun(Off) ->
		GuessUp = FirstGuess + Off,
		Answer = xmr_test(GuessUp),
		if
			Answer == false -> [];
			true -> [GuessUp]
		end
	      end, lists:seq(1, ThreadNum)).


get_prime_list_main(_, [], _, PrimeList) ->
	PrimeList;
get_prime_list_main(FirstGuess, [Start | End], ThreadNum, PrimeList) ->
	GuessUp = FirstGuess + (Start * ThreadNum),
	PrimeList0 = get_prime_list(GuessUp, ThreadNum) ++ PrimeList,
	get_prime_list_main(FirstGuess, End, ThreadNum, PrimeList0).


upmap(F, L) ->
	Parent = self(),
	Ref = make_ref(),
	[receive {Ref, Result} ->
		Result
	 end
		|| _ <- [spawn(fun() -> Parent ! {Ref, F(X)} end) || X <- L]].












	

File: maxprime.go
package main

import (
	"fmt"
	//"os"
	"math"
)

func main() {
	fmt.Println("Please enter the number:")
	var Input int64
	fmt.Scanf("%d", &Input)
	Maxprime := get_maxprime(Input)
	fmt.Println("The max prime is:")
	fmt.Println(Maxprime)
}

func get_maxprime(Input int64) (Maxprime int64) {
	var Is_prime bool = xmr_Test(Input)
	if Is_prime == true {
		Maxprime = Input
		return
	} else {
		Maxprime = guess_maxprime(Input)
		return
	}
}

func guess_maxprime(Input int64) (MaxPrime int64) {
	var Guess int64 = first_guess(Input)
	if Guess == Input {
		MaxPrime = down_guess(Input)
		return
	} else {
		var Up_guess int64 = up_guess(Input, Guess)
		if Up_guess == Guess {
			var Is_prime bool = xmr_Test(Guess)
			if Is_prime == true {
				MaxPrime = Guess
				return
			} else {
				MaxPrime = down_guess(Input)
				return
			}
		} else {
			MaxPrime = Up_guess
			return
		}
	}
}

func xmr_Test(Input int64) bool {
	if Input < 2047 {
		return mrTest(2, Input)
	} else {
		return mrTest(2, Input) &&
			mrTest(7, Input) &&
			mrTest(61, Input)
	}
}

func mrTest(baseNum int64, Input int64) bool {
	if evenOrNot(Input) == false {
		var minus1 int64 = Input - 1
		var modMon int64 = mon(baseNum, minus1, Input)
		if modMon == 1 {
			return mrTestMain(baseNum, Input, minus1)
		} else {
			return false
		}
	} else {
		return false
	}
}

func mrTestMain(baseNum int64, Input int64, minus1 int64) (passOrNot bool) {
	var half int64 = minus1 % 2
	var halfmod int64 = mon(baseNum, half, Input)
	var i, j, k int64
	for i, j, k = 0, half, halfmod; i >= 0; i++ {
		if evenOrNot(i) == false {
			if (halfmod == (Input - 1)) || halfmod == 1 {
				passOrNot = true
				return
			} else {
				passOrNot = false
				return
			}
		} else {
			if k == (Input - 1) {
				passOrNot = true
				return
			} else if k != 1 {
				passOrNot = false
				return
			} else {
				j = j % 2
				k = mon(baseNum, j, Input)
			}
		}
	}
	return passOrNot
}

func evenOrNot(num int64) bool {
	return (num & 1) == 0
}

func mon(baseNum int64, minus1 int64, Input int64) int64 {
	var modBaseNum int64 = baseNum % Input
	var modMon int64 = 1
	mod := monMain(minus1, modMon, modBaseNum, Input)
	return mod
}

func monMain(minus1 int64, modMon int64, modBaseNum int64, Input int64) int64 {
	var numLeft int64 = minus1
	var modFinal int64 = modMon
	j := modFinal
	for i , k := numLeft, modBaseNum; i != 0; i, k = i/2, (k*k)%Input {
		if evenOrNot(i) != true {
			j = (j * k) % Input
		}

	}
	return j % Input
}

func down_guess(firstGuess int64) int64 {
	maxprime := firstGuess - 1
	for i := maxprime; xmr_Test(i) == false; i-- {
		maxprime--
	}
	return maxprime
}

func up_guess(input int64, firstGuess int64) int64 {
	maxprime := firstGuess
	for i := maxprime; i < input; i++ {
		if xmr_Test(i) == true {
			maxprime = i
		}
	}
	return maxprime
}

func first_guess(Input int64) (Guess int64) {
	var Gap float64 = gap(Input)
	Guess = Input - int64(Gap+0.5)
	return
}

func gap(Input int64) float64 {
	Inputf := float64(Input)
	Inputln := math.Log(Inputf)
	var Gap float64 = Inputln /
		(1 +
			(1 / Inputln) +
			(2 / (math.Pow(Inputln, 2))) +
			(6 / (math.Pow(Inputln, 3))) +
			(24 / (math.Pow(Inputln, 4))) +
			(120 / (math.Pow(Inputln, 5))) +
			(720 / (math.Pow(Inputln, 6))) +
			(5040 / (math.Pow(Inputln, 7))) +
			(40320 / (math.Pow(Inputln, 8))))
	return Gap
}
